transforms.md

```diff
@@ -0,0 +1,167 @@
+
+    @staticmethod
+    def image(input, *, bar, baz):
+        return foo(input, bar, baz=baz)
+```
+
+1. The new style is a little more verbose. This is due to the fact that `Transform`'s are now able to handle multiple input types and a single source of parameters is needed for them. For random transforms the verbosity stays the same, since they already have this single source for the random parameters.
+2. If `foo` only takes one positional argument, defining a custom `image` method could also be replaced by `self.register_feature_transform(Image, foo)` in `__init__()`.
+
+## Random transforms
+
+There are in general two types of random transforms:
+
+1. Transforms that sample their parameters at random for each sample, but are always applied.
+2. Transforms that are applied at random given a probability, but have fixed parameters.
+
+These cases are not mutually exclusive, but since they address independent concepts they also can be handled independently.
+
+### Random parameters
+
+Since the dispatch for the complete sample happens from a single point in the `forward` method, it is easy to perform the transforms with the same random parameters. For example, for `RandomRotate` the implementation might look like
+
+   ```python
+   class RandomRotate(Transform):
+       def __init__(self, low: float, high: float) -> None:
+           super().__init__()
+           self._dist = torch.distributions.Uniform(low, high)
+   
+       def get_params(self, sample: Any = None) -> Dict[str, Any]:
+           return dict(degrees=self._dist.sample().item())
+   
+       @staticmethod
+       def image(image: Image, *, degrees: float) -> Image:
+           return image
+   
+       @staticmethod
+       def bounding_box(bounding_box: BoundingBox, *, degrees: float) -> BoundingBox:
+           return bounding_box
+   ```
+
+   Since we can optionally pass the `sample` to `.get_params()`, the parameters can also depend on the input, which is required for some transforms like [`RandomErasing`](https://github.com/pytorch/vision/blob/f3aff2fa60c110c25df671b6f99ffb26727cb8ae/torchvision/transforms/transforms.py#L1608). Additionally, if required, we can also use the same set of random parameters for multiple distinct transforms:
+
+   ```python
+   params = transform1.get_params()
+   output1 = transform1(input1, params=params)
+   output2 = transform2(input2, params=params)
+   ```
+
+### Random application
+
+Since the randomness only decides if the transformation is applied or not, this could be handled well in the dispatch mechanism. By adding a `p=1.0` or `probability=1.0` keyword argument to `Transform`, the dispatch logic can decide if the transformation is applied or not without user interference. This has the advantage of combining a "standard" and random transform that otherwise do almost the same. For example `RandomHorizontalFlip(p=0.5)` would simply become `HorizontalFlip(p=0.5)`.
+
+Let me know if you think this is a worthwhile addition.
```

Comment:
Why not having a class
```python
class HorizontalFlip(RandomHorizontalFlip):
    def __init__(self):
        super().__init__(p=1.0)
```
That would alleviate the need of coding up twice the same thing but keep the nice UX of having access to a class that is not random.
Would also make scripts clearer.
There is also another way of doing things, which is to have a class `HorizontalFlip` that does the job and a class RandomHorizontalFlip that would just call HorizontalFlip with probability `p` (in other words, it is the opposite version of what i just suggested).
That would also mean that we could have a code that is recycled across transforms.

Something like
```python
class HorizontalFlip(Transform):
    @staticmethod
    def apply_image(image):
        return image.hflip()
```

and 
```python

def random_apply(fun):
    def decorated_fun(self, image, **params):
        p = self.p
        if torch.rand(1).item() > p:
            return fun(self, image, **params)
        return image
    return decorated_fun


class RandomHorizontalFlip(Transform):
    def __init__(self, p):
        super().__init__()
        self.p = p

    @random_apply
    def apply_image(self, image):
        return super().apply_image(image)
```

Reply:
> That would alleviate the need of coding up twice the same thing but keep the nice UX of having access to a class that is not random.

The only issue I see with this is that we need to maintain two sets of docstrings. On the other hand, if we were to apply this pattern to multiple transforms, there is probably some room for automation in this. 

> Would also make scripts clearer.

Yeah, that is a valid concern. Another idea would be to have a `RandomApply` object that wraps a transform, i.e. `transform = RandomApply(Rotate(), p=0.5)`. That should be as clear as having a separate class, but would drastically reduce the maintenance burden.